### DISCUSSION TITLE
Leveraging Automatic's new API `alwayson_scripts` field

### DISCUSSION THREAD

**Discussion by Hugo-Matias** [adieyal/sd-dynamic-prompts#308] [General]
The WebUI API was recently updated to allow script arguments to be passed in to `alwayson_scripts`, if I'm not mistaken dynamic-prompts fall under that category right? 

Is there a way we can now consume the script from an external source via the api endpoints? Things like Magic Prompt would be really interesting to implement. With proper documentation about the correct argument variables to send we could make it work.

**Comment by adieyal** [comment]
I haven't looked at the updated script arguments yet, in the meanwhile, have you had a look at [dynamicprompts](https://github.com/adieyal/dynamicprompts)? That's the library that drives the extension. Depends on what you're  trying to do, it might be more convenient than using Auto1111's api.

**Reply by Hugo-Matias** [reply]
Thanks for the reply @adieyal, the guys at the [ControlNet extension](https://github.com/Mikubill/sd-webui-controlnet) are currently migrating their custom api to the new argument based call, check them out if you need some tips.

I'm consuming the api from a Blazor server app using C# and I was actually looking at [pythonnet](https://pypi.org/project/pythonnet/) to implement the library but I'm still not sure if that's the best approach going forward, it would add complexity to the project but could provide more flexibility and it's still an opportunity to challenge myself with new problems.

**Comment by Hugo-Matias** [comment]
Quick update on this. Looks like the `alwayson_scripts` implementation allows for out of the box integration of the scripts without much hassle. The arguments passed into the [`process` function](https://github.com/adieyal/sd-dynamic-prompts/blob/1596f556e23be4865206a0d3c1c59451a17d506a/sd_dynamic_prompts/dynamic_prompting.py#L327) of the script are the ones used in the `args` field of the payload, with that, it's just a matter of listing them in the proper order and it will most likely work. I was able to implement Dynamic Prompts and Cutoff into my workflow quite quickly (ControlNet as well but they use an in-house solution that hijacks the api call in order to be able to receive the control net units as a list of dictionaries instead). As an example here is a working payload (remove comments to try it out):

```json
 {
  "enable_hr": false,
  "prompt": "",
  "seed": -1,
  "sampler_name": "DDIM",
  "batch_size": 1,
  "n_iter": 1,
  "steps": 20,
  "cfg_scale": 7,
  "width": 512,
  "height": 512,
  "alwayson_scripts": {
	"Dynamic Prompts v2.8.4": {  // def title():
		"args": [
			true,  // is_enabled: bool
			false, // is_combinatorial: bool
			1,  // combinatorial_batches: int
			true,  // is_magic_prompt: bool
			true, // is_feeling_lucky: bool
			false,  // is_attention_grabber: bool
			1.1,  // min_attention: float
			1.5,  // max_attention: float
			100,  // magic_prompt_length: int
			0.7,  // magic_temp_value (I believe this is magic prompt creativity): float
			false,  // use_fixed_seed: bool
			false,  // unlink_seed_from_prompt: bool
			false,  // disable_negative_prompt: bool
			false,  // enable_jinja_templates: bool
			false,  // no_image_generation: bool
			0,  // max_generations: int
			"Gustavosta/MagicPrompt-Stable-Diffusion",  // magic_model: string
			null  // magic_blocklist_regex: string or null
		]
	}
  }
}
```

Has you might have noticed the dictionary key must be the [script `title`](https://github.com/adieyal/sd-dynamic-prompts/blob/1596f556e23be4865206a0d3c1c59451a17d506a/sd_dynamic_prompts/dynamic_prompting.py#L119). Unless we somehow parse the script file at runtime, it will break the implementation when the value changes, it might be a nuisance for some but it perhaps is for the best and makes sure everyone updates their code.

The arguments are the ones passed into the `process` function and must be sent in the order they are consumed.

**Reply by Piprian** [reply]
Thank you for this!
I can't seem to get this to work. Has this been changed?
If I understand the error messages correctly, it seems I am not getting any images back whenever I try to add dynamic prompts to my API requests.

I have no idea what I am doing so I apologize if my questions are stupid.

Edit in case anybody else has the same problem:
It seems you need to include the version number when calling the script. I thought it had to be the name of the extension's folder (like it is with the one other extension I have used) which in this case is just "Dynamic Prompts".
"Dynamic Prompts v2.12.5" works.

**Reply by Piprian** [reply]
Has there been any progress on this? I can't find anything.
It's a little annoying to have to change my script every time I update dynamic prompts.
